src/agent/diagnostics.ts

```diff
@@ -44,10 +43,128 @@ export class DiagnosticFileWriter implements cm.IDiagnosticWriter  {
     divider() {
         this.write('----------------------------------------');
     }
+    
+    public end(): void {
+    }               
+}
 
+export class RollingDiagnosticFileWriter implements cm.IDiagnosticWriter {
+    constructor(level: cm.DiagnosticLevel, folder: string, filenamePrefix: string, settings: cm.ILogSettings) {
+        this.level = level;
+        this._folder = folder;
+        this._filenamePrefix = filenamePrefix;
+        this._maxLinesPerFile = settings.linesPerFile;
+        this._filesToKeep = settings.maxFiles;
+        
+        this._initializeFileQueue();
+    }
+    
+    public level: cm.DiagnosticLevel;
+    private _folder: string;
+    private _filenamePrefix: string;
+    private _fileDescriptor: number;
+    private _lineCount: number = 0;
+    private _maxLinesPerFile: number;
+    private _filesToKeep: number;
+    private _fileQueue: string[];
+    private _previouslyGeneratedFilename: string;
+    private _uniqueFileCounter: number = 1;
+    
+    public write(message: string): void {
+        var fileDescriptor = this._getFileDescriptor();
+        fs.writeSync(fileDescriptor, message);
+        if (message) {
+            // count newlines
+            this._lineCount += message.split('\n').length - 1;
+        }
+    }
+    
+    public writeError(message: string): void {
+        this.write(message);
+    }
+    
     public end(): void {
+        if (this._fileDescriptor) {
+            fs.closeSync(this._fileDescriptor);
+        }
+    }
+    
+    private _getFileDescriptor(): number {
+        if (this._fileDescriptor && this._lineCount >= this._maxLinesPerFile) {
+            // close the current file
+            fs.closeSync(this._fileDescriptor);
+            this._fileDescriptor = undefined;
+        }
         
-    }               
+        if (!this._fileDescriptor) {
+            // create a new file and reset the line count
+            var filename = this._generateFilename();
+            this._lineCount = 0;
+            this._fileDescriptor = fs.openSync(filename, 'a');
+            
+            // add the filename to the queue and delete any old ones
+            this._fileQueue.push(filename);
+            while (this._fileQueue.length > this._filesToKeep) {
+                fs.unlinkSync(this._fileQueue.splice(0, 1)[0]);
```

Comment:
We should only have async I/O for things in the worker to ensure we don't block all the queues etc... that do I/O.  Individually not a big deal but in aggregate would be a problem if a bunch of worker code started sync method usages.  We should only allow sync in config, agent (it's just a message loop) and tasks (they run in a separate single purpose process)